wgpu-core/src/macros.rs

```diff
@@ -2,140 +2,56 @@
-
-        #[cfg(windows)]
-        let $dx11_field = $dx11_expr;
-    };
-
-    // one block statement per backend
-    (
-        #[vulkan] $vk_block:block
-        #[metal] $mtl_block:block
-        #[dx12] $dx12_block:block
-        #[dx11] $dx11_block:block
-    ) => {
-        #[cfg(any(
-            windows,
-            all(unix, not(any(target_os = "ios", target_os = "macos"))),
-            feature = "gfx-backend-vulkan",
-        ))]
-        $vk_block
-
-        #[cfg(any(target_os = "ios", target_os = "macos"))]
-        $mtl_block
-
-        #[cfg(windows)]
-        $dx12_block
-
-        #[cfg(windows)]
-        $dx11_block
-    };
-
-    // a struct constructor with one field per backend
-    (
-        $Struct:path {
-            #[vulkan] $vk_field:ident: $vk_expr:expr,
-            #[metal] $mtl_field:ident: $mtl_expr:expr,
-            #[dx12] $dx12_field:ident: $dx12_expr:expr,
-            #[dx11] $dx11_field:ident: $dx11_expr:expr,
-        }
-    ) => {{
-        $Struct {
-            #[cfg(any(
-                windows,
-                all(unix, not(any(target_os = "ios", target_os = "macos"))),
-                feature = "gfx-backend-vulkan",
-            ))]
-            $vk_field: $vk_expr,
-
-            #[cfg(any(target_os = "ios", target_os = "macos"))]
-            $mtl_field: $mtl_expr,
-
-            #[cfg(windows)]
-            $dx12_field: $dx12_expr,
-
-            #[cfg(windows)]
-            $dx11_field: $dx11_expr,
-        }
-    }};
-}
-
 macro_rules! backends_map {
     // one let statement per backend with mapped data
     (
         let map = |$backend:pat| $map:block;
         $(
-            #[$backend_attr:ident] let $pat:pat = map($expr:expr);
+            #[cfg($backend_cfg:meta)] let $pat:pat = map($expr:expr);
         )*
     ) => {
-        backends! {
-            $(
-                #[$backend_attr]
-                let $pat = {
-                    let $backend = $expr;
-                    $map
-                };
-            )*
-        }
+        $(
+            #[cfg($backend_cfg)]
+            let $pat = {
+                let $backend = $expr;
+                $map
+            };
+        )*
     };
 
     // one block statement per backend with mapped data
     (
         let map = |$backend:pat| $map:block;
         $(
-            #[$backend_attr:ident] map($expr:expr),
+            #[cfg($backend_cfg:meta)] map($expr:expr),
         )*
     ) => {
-        backends! {
-            $(
-                #[$backend_attr]
-                {
-                    let $backend = $expr;
-                    $map
-                }
-            )*
-        }
+        $(
+            #[cfg($backend_cfg)]
+            {
+                let $backend = $expr;
+                $map
+            }
+        )*
     };
 
     // a struct constructor with one field per backend with mapped data
     (
         let map = |$backend:pat| $map:block;
-        $Struct:path {
+        $Struct:ident {
```

Comment:
For reasons I cannot yet fathom, the macro will not work in the `cargo test` anymore unless this is switched to `ident`.  I think this will still work in most cases and the tests pass after this change. I'm not sure if this is a problem or not.